src/servers/src/error.rs

```diff
@@ -752,15 +766,13 @@ impl IntoResponse for Error {
             | Error::InvalidPromRemoteRequest { .. }
             | Error::InvalidQuery { .. }
             | Error::TimePrecision { .. } => HttpStatusCode::BAD_REQUEST,
-            _ => {
-                if self.status_code().should_log_error() {
-                    error!(self; "Failed to handle HTTP request: ");
-                } else {
-                    debug!("Failed to handle HTTP request: {self:?}");
-                }
-
-                HttpStatusCode::INTERNAL_SERVER_ERROR
-            }
+            Error::Pipeline { ref source, .. } => match source {
+                PipelineError::CompilePipeline { .. }
+                | PipelineError::InvalidPipelineVersion { .. }
+                | PipelineError::PipelineNotFound { .. } => HttpStatusCode::BAD_REQUEST,
+                _ => log_error_and_return_internal_server_error_http_status(&self),
+            },
+            _ => log_error_and_return_internal_server_error_http_status(&self),
```

Comment:
I think it's better to avoid side-effect here because in some case we may forget to use this log function and just return the INTERNAL_SERVER_ERROR here.

We can simply call the `log_error_if_necessary` after this pattern matching.

Reply:
Agree.